constructors.md

```diff
@@ -1 +1,112 @@
+ * data and bitmask that will be *deep* copied.
+ *
+ * @param[in] dtype The element type
+ * @param[in] data_buffer device_buffer whose data will be *deep* copied
+ * @param[in] mask_buffer Optional device_buffer whose data will be *deep*
+ *copied
+ *---------------------------------------------------------------------------**/
+column(DType dtype, device_buffer data,
+       device_buffer mask_buffer = device_buffer{});
+
+/**---------------------------------------------------------------------------*
+ * @brief Construct a new column from a type, and device_buffers for data and
+ * bitmask that will be *shallow* copied.
+ *
+ * This constructor uses move semantics to take ownership of the device_buffer's
+ * device memory. The `device_buffer` passed into this constructor will not
+ * longer be valid to use. Furthermore, it will result in undefined behavior if
+ * the device_buffer`s associated memory is modified or freed after invoking
+ * this constructor.
+ *
+ * @param dtype The element type
+ * @param data device_buffer whose data will be moved from into this column
+ * @param mask Optional device_buffer whose data will be moved from into this
+ * column. If no device_buffer is passed in, it is assumed all elements are
+ * valid and no bitmask is allocated.
+ *---------------------------------------------------------------------------**/
+column(DType dtype, device_buffer&& data,
+       device_buffer&& mask_buffer = device_buffer{});
+
+/**---------------------------------------------------------------------------*
+ * @brief Construct a new column from a type, and a device_buffer for data.
+ *
+ * @note If `dtype` is a floating point type, an associated bitmask will be
+ * created where every NaN value is set to NULL. Otherwise, the bitmask will not
+ * be allocated.
+ *
+ * @param dtype The element type
+ * @param data device_buffer whose data will be deep copied into this column
+ * @param nan_as_null If `dtype` is a floating point type, optionally allocate a
+ * bitmask where every NaN value is set to NULL.
+ *---------------------------------------------------------------------------**/
+column(DType dtype, device_buffer data, bool nan_as_null = true);
+
+/**---------------------------------------------------------------------------*
+ * @brief Construct a new column from a type, and a device_buffer whose device
+ * memory will be moved into this column.
+ *
+ * @note If `dtype` is a floating point type, an associated bitmask will be
+ * created where every NaN value is set to NULL. Otherwise, the bitmask will not
+ * be allocated.
+ *
+ * @param dtype The element type
+ * @param data device_buffer whose data will be moved into this column
+ * @param nan_as_null If `dtype` is a floating point type, optionally allocate a
+ * bitmask where every NaN value is set to NULL.
+ *---------------------------------------------------------------------------**/
+column(DType dtype, device_buffer&& data, bool nan_as_null = true);
```

Comment:
I was wondering if the nans-as-nulls behavior would be better exposed as a factory function (`from_pandas_data` or whatever) that creates the validity buffer from the data values if necessary then calls the column constructor.  That would eliminate the need for the nan_as_null optional parameter since calling that factory method means nans-as-nulls is what they want.

Reply:
Great point! That's what I get for writing code in Markdown. 

How about instead we do:

```
column(DType dtype, device_buffer&& data, bool allocate_bitmask = false, bool nan_as_null = true);
```

That improves this API as it gives you the choice on whether or not the bitmask is allocated. Otherwise, as it was before, a bitmask would _always_ be allocated for floating point types.

Reply:
@jlowe I came to the same conclusion. @revans2 see updated constructors that are much more stripped down.

The `nan_as_nulls` functionality was originally requested by @kkraus14, but it can be emulated by doing something like:

```
float* raw_float_data = ... ;

device_buffer float_buffer = device_buffer(raw_float_data, size);

// create a column w/ no bitmask from a device buffer of floats
cudf::column float_col(GDF_FLOAT32, size, float_buffer);

// create a bitmask where every NaN is null
bitmask = cudf::nans_to_nulls(float_col);

// construct a new column by shallow copying the original columns data and the new bitmask
cudf::column float_col_with_bitmask( GDF_FLOAT32, size, std::move(float_col.data()), std::move(bitmask));
```